include/picongpu/particles/debyeLength/Check.kernel

```diff
@@ -0,0 +1,227 @@
+                            check(acc, electronBox, supercellIdx, variance, result);
+                        });
+                    }
+
+                    /** Check Debye length and write back results
+                     *
+                     * This function makes an estimate for the whole provided electron box,
+                     * and so must be called once per non-empty supercell.
+                     *
+                     * @tparam T_Acc alpaka accelerator type
+                     * @tparam T_ElectronBox electron storage type
+                     * @tparam T_ResultBox result storage type, device box for Result
+                     *
+                     * @param acc alpaka accelerator
+                     * @param electronBox electron storage
+                     * @param supercellIdx supercell index, including guards
+                     * @param momentumVariance estimated variance in physical momentums
+                     * @param result device box of type Result with a single element
+                     *        to write kernel result to
+                     */
+                    template<typename T_Acc, typename T_ElectronBox, typename T_ResultBox>
+                    DINLINE void check(
+                        T_Acc const& acc,
+                        T_ElectronBox const electronBox,
+                        pmacc::DataSpace<simDim> const supercellIdx,
+                        double momentumVariance,
+                        T_ResultBox result) const
+                    {
+                        auto numParticles
+                            = static_cast<float_64>(electronBox.getSuperCell(supercellIdx).getNumParticles());
+                        auto const supercellVolume
+                            = CELL_VOLUME * precisionCast<float_64>(SuperCellSize::toRT().productOfComponents());
+                        auto const electronDensity = numParticles / supercellVolume;
+
+                        /* For the Maxwell-Boltzmann distribution,
+                         * variance = mass * k * T, both for physical and macroparticle.
+                         * We assume this distribution and estimate k * T out of variance
+                         */
+                        using Frame = typename T_ElectronBox::FrameType;
+                        auto const mass = static_cast<double>(picongpu::traits::frame::getMass<Frame>());
+                        auto const kT = momentumVariance / mass;
+                        auto const electronCharge = picongpu::traits::frame::getCharge<Frame>();
+                        auto const debyeLength = sqrt(EPS0 * kT / (electronDensity * electronCharge * electronCharge));
+
+                        auto maxCellSize = cellSize[0];
+                        for(uint32_t d = 1; d < simDim; d++)
+                            maxCellSize = math::max(maxCellSize, cellSize[d]);
+                        auto const minCellsPerDebyeLength = 2.0;
```

Comment:
By the way, Hockney says `minCellsPerDebyeLength = 1.0`, while Birdsall provides results of numerical experiments for `minCellsPerDebyeLength = 3.0` :man_shrugging: